File: app/src/main/java/tpcreative/co/qrscanner/common/SettingsSingleton.java
package tpcreative.co.qrscanner.common;

public class SettingsSingleton {

    private static SettingsSingleton instance ;
    private SettingsSingleton.SingletonSettingsListener listener;

    public static SettingsSingleton getInstance(){
        if (instance==null){
            synchronized (SettingsSingleton.class){
                if (instance==null){
                    instance = new SettingsSingleton();
                }
            }
        }
        return instance;
    }

    public void setListener(SettingsSingleton.SingletonSettingsListener listener){
        this.listener = listener;
    }

    public void onUpdated(){
        if (listener!=null){
            listener.onUpdated();
        }
    }

    public void onUpdateSharePreference(boolean value){
        if (listener!=null){
            listener.onUpdatedSharePreferences(value);
        }
    }

    public interface SingletonSettingsListener{
        void onUpdated();
        void onUpdatedSharePreferences(boolean value);
    }
}

File: app/src/main/java/tpcreative/co/qrscanner/common/SettingsSingleton.kt
package tpcreative.co.qrscanner.common

class SettingsSingleton {
    private var listener: SingletonSettingsListener? = null
    fun setListener(listener: SingletonSettingsListener?) {
        this.listener = listener
    }

    fun onUpdated() {
        if (listener != null) {
            listener?.onUpdated()
        }
    }

    fun onSyncDataRequest() {
        if (listener != null) {
            listener?.onSyncDataRequest()
        }
    }

    fun onUpdateSharePreference(value: Boolean) {
        if (listener != null) {
            listener?.onUpdatedSharePreferences(value)
        }
    }

    fun onUpdatedPremiumVersion(){
        if (listener != null){
            listener?.onUpdatedPremiumVersion()
        }
    }

    interface SingletonSettingsListener {
        fun onUpdated()
        fun onUpdatedSharePreferences(value: Boolean)
        fun onSyncDataRequest()
        fun onUpdatedPremiumVersion()
    }

    companion object {
        private var instance: SettingsSingleton? = null
        fun getInstance(): SettingsSingleton? {
            if (instance == null) {
                synchronized(SettingsSingleton::class.java) {
                    if (instance == null) {
                        instance = SettingsSingleton()
                    }
                }
            }
            return instance
        }
    }
}